module/modmanager/manager.go

```diff
@@ -195,9 +195,10 @@ func (mgr *Manager) Handles() map[string]modlib.HandlerMap {
 //
 // Each module configuration should have a unique name - if duplicate names are detected,
 // then only the first duplicate instance will be processed and the rest will be ignored.
-//
-// This method is not thread-safe.
 func (mgr *Manager) Add(ctx context.Context, confs ...config.Module) error {
+	mgr.mu.Lock()
+	defer mgr.mu.Unlock()
```

Comment:
Would you mind walking me through how this doesn't introduce deadlock with any attempted restarts offline?

Reply:
my understanding is that this would behave similarly to when `Add` was sequential - in that version of module manager, `Add` and `attemptRestart` both grabbed the manager lock without deadlocking - `Add` was able to finish and `attemptRestart` would run afterwards.

the difference now is that multiple `attemptRestart`s run if multiple modules crash while `Add`ing. my understanding is that this is fine - the `attemptRestart`s would run sequentially in some order after the batch `Add` is complete. each module has an `inRecovery` lock that prevents multiple `attemptedRestart`s from queuing up for the same module, but it's okay for restarts to queue up for different modules.

so the flow would be something like:

* `Add(mod1, mod2, mod3)`
* `mod1` succeeds but `mod2` and `mod3` crash
* `mod2` and `mod3` call their `attemptedRestart` functions, which idle until `mod1` is added
* the two `attemptedRestart` functions will be executed in some order sequentially

all this said, it's worth adding a testing that demonstrates the flow I am describing - i.e concurrently add two modules that both crash and show that there is no deadlock. I'll add that as part of this PR.

Reply:
> the difference now is that multiple attemptRestarts run if multiple modules crash while Adding. my understanding is that this is fine - the attemptRestarts would run sequentially in some order after the batch Add is complete. each module has an inRecovery lock that prevents multiple attemptedRestarts from queuing up for the same module, but it's okay for restarts to queue up for different modules.

Also my understanding from reading the code, and thank you for writing it out! I think adding a test as you've described and running it with `-race` should be enough to say that this `Lock` you're adding is not _creating_ any deadlock or races.

Reply:
ah, so [I added a test for this scenario](https://github.com/viamrobotics/rdk/pull/3701/commits/39f249a3edd0cb5057243741367da2a99af18a4d) and do see a race condition! something about my assumptions we off (or I wrote the test poorly)